### ISSUE TITLE
CA1859 false positive with arrays

### ISSUE THREAD

**Issue by JustArchi** [dotnet/roslyn-analyzers#6931]
### Analyzer

**Diagnostic ID**: [CA1859](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1859): Use concrete types when possible for improved performance

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 8 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 8.0.100-rc1](https://dotnet.microsoft.com/download/dotnet/8.0)

### Describe the bug

False positive of CA1859 where I don't expect one.

### Steps To Reproduce

I got false positive with following code:

```csharp
	public static async Task<IList<T>> InParallel<T>(IEnumerable<Task<T>> tasks) {
		ArgumentNullException.ThrowIfNull(tasks);

		IList<T> results;

		switch (ASF.GlobalConfig?.OptimizationMode) {
			case GlobalConfig.EOptimizationMode.MinMemoryUsage:
				results = new List<T>();

				foreach (Task<T> task in tasks) {
					results.Add(await task.ConfigureAwait(false));
				}

				break;
			default:
				results = await Task.WhenAll(tasks).ConfigureAwait(false);

				break;
		}

		return results;
	}
```

Rewritten for easier reproduction to the code that compiles:

```csharp
	public static async Task<IList<T>> InParallel2<T>(IEnumerable<Task<T>> tasks) {
		ArgumentNullException.ThrowIfNull(tasks);

		IList<T> results;

		switch (DateTime.Now.Year) {
			case 2023:
				results = new List<T>();

				foreach (Task<T> task in tasks) {
					results.Add(await task.ConfigureAwait(false));
				}

				break;
			default:
				results = await Task.WhenAll(tasks).ConfigureAwait(false);

				break;
		}

		return results;
	}
```

### Expected behavior

No CA1859 in above code.

### Actual behavior

Here roslyn analyzer suggests me to change `IList<T> results` into `List<T>`, but doesn't take into account that `Task.WhenAll()` returns `T[]` which is not a `List<T>`.

## Additional context

I reproduced this with clean .NET 8 rc1 which was released yesterday so I believe this issue is recent - please feel free to reproduce and see if it applies.

**Comment by pcf0** [comment]
I think this is the same problem as in my case? #6904 (value is changed by await)
